Allow quantumcircuit construction with zero classical registers

peendebak opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

Allow circuits with zero classical registers to be created in the same way as circuits with classical registers.


### Details and comments

An example
```
from qiskit.circuit import QuantumCircuit
c = QuantumCircuit(2,2)
c = QuantumCircuit(2,0) # was not possible
```

mtreinish commented:
So I'm curious is there an advantage to doing this over:
 
```python
from qiskit.circuit import QuantumCircuit
c = QuantumCircuit(2)
```

because that works fine today. Adding support for 0 seems kind of weird to me because we enforce a positive non-zeo int for qubits.

Cryoris commented:
I could see the value of allowing 0 bits in the initializer, e.g. if you automatically construct and compose circuits and some of them happen to have 0 qubits/clbits. But I agree with @mtreinish, it's a bit strange to only have that for classical bits. If we add this, I think we should have this for both arguments and also for the registers (i.e. `QuantumRegister(0)`).

peendebak commented:
@mtreinish The case that @Cryoris describes is exactly what we encountered and why this PR was made. 

I updated the PR so that quantum register of size 0 are allows, and I updated the tests.

ajavadia commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -1151,7 +1151,10 @@ def add_register(self, *regs):
             elif len(regs) == 2 and all(isinstance(reg, int) for reg in regs):
                 # QuantumCircuit with anonymous wires e.g. QuantumCircuit(2, 3)
-                regs = (QuantumRegister(regs[0], "q"), ClassicalRegister(regs[1], "c"))
+                if regs[1] == 0:
+                    regs = (QuantumRegister(regs[0], "q"),)
+                else:
+                    regs = (QuantumRegister(regs[0], "q"), ClassicalRegister(regs[1], "c"))
```
If size 0 is allowed then this modification is not needed i think

eendebakpt replied:
@ajavadia Removing this `if` statement triggers some issues with the tests. There is a difference between a circuit with no classical registers (i.e. an empty list), or a circuit with a classical register of size 0.

![image](https://user-images.githubusercontent.com/883786/122742366-27327180-d286-11eb-9a87-93f9f56329a1.png)

The behaviour in this PR. e.g. `QuantumCircuit(2,0)` creates a circuit without classical registers, is consistent with current behaviour (and tests).

ajavadia commented on qiskit/circuit/register.py:
```diff
@@ -82,7 +82,7 @@ def __init__(self, size=None, name=None, bits=None):
         size = int(size)  # cast to int
 
-        if size <= 0:
+        if size < 0:
             raise CircuitError(
                 "Register size must be positive (%s '%s' was provided)"
```
the error should now say "non-negative"

eendebakpt replied:
@ajavadia As a mathematician positive includes zero (and strictly positive does not include zero). But the term non-negative is clearer indeed.

ajavadia reviewed (commented):
I think it makes sense to allow zero-length registers in general (they're basically like an empty list of Qubits or empty list of Clbits). I certainly prefer this than the stop-gap solution of allowing `QuantumCircuit(n, 0)` to work.

But we need to make sure the rest of Qiskit is prepared to handle circuits circuits with zero-length registers. For example the circuit drawers or the transpiler should not throw errors.

peendebak commented:
@mtreinish Could you have another look at the PR?

peendebak commented:
@ajavadia Could you have a look at the PR and let me know whether it needs some more improvements?

kdk reviewed (commented):
Code changes so far LGTM, but I agree with @mtreinish and @Cryoris that this introduces an unusual difference in behavior between `QuantumRegister`s and `ClassicalRegister`s.

After this PR, `QuantumCircuit(0, 0)` would have one 0-length `QuantumRegister`, but no `ClassicalRegister`s. I'd vote for also allowing `QuantumCircuit(0)` and `QuantumCircuit(0, N)` to not create a `QuantumRegister`.

eendebakpt commented:
@mtreinish @Cryoris @kdk The current behaviour is
![image](https://user-images.githubusercontent.com/883786/125802549-0b7c51a7-02a8-4192-a213-e55ea43f1684.png)

There is indeed a difference in behaviour between the registers, but it is most consistent with the current behaviour. I don't mind changing to another type of behaviour, but please first come to consensus before I update the PR again.

eendebakpt commented:
@kdk The behavior you suggested I implemented in #6800

kdk commented:
Closing in favor of #6800 .
